Value cannot be null. Parameter name: type

Release 6.2.0 on dnn 8.0.3.
It happens randomly, the form and all other html modules on the page look blank; page refresh does not help. One has to go to a different page and back to see the form and the rest of the page working again. Here is the log message:

AbsoluteURL:/Default.aspx
DefaultDataProvider:DotNetNuke.Data.SqlDataProvider, DotNetNuke
ExceptionGUID:7c7b50e8-eaf3-4c3b-8ca5-4a66b01c28b3
AssemblyVersion:
PortalId:-1
UserId:-1
TabId:-1
RawUrl:
Referrer:
UserAgent:
ExceptionHash:uQ+8ZRsItfjnKYfnr2T38g==
Message:Value cannot be null. Parameter name: type
StackTrace:
   at System.Activator.CreateInstance(Type type, Boolean nonPublic)
   at System.Activator.CreateInstance(Type type)
   at DotNetNuke.Modules.UserDefinedTable.DataType.get_DataTypes()
   at DotNetNuke.Modules.UserDefinedTable.DataType.ByName(String dataTypeName)
   at DotNetNuke.Modules.UserDefinedTable.EditForm.BuildEditForm()
   at DotNetNuke.Modules.UserDefinedTable.EditForm.OnInit(EventArgs e)
   at System.Web.UI.Control.InitRecursive(Control namingContainer)
   at System.Web.UI.Control.AddedControl(Control control, Int32 index)
   at System.Web.UI.ControlCollection.Add(Control child)
   at DotNetNuke.Modules.UserDefinedTable.Default.LoadControlByKey(String controlKey)
   at DotNetNuke.Modules.UserDefinedTable.Default.InitViews()
   at DotNetNuke.Modules.UserDefinedTable.Default.OnInit(EventArgs e)
   at System.Web.UI.Control.InitRecursive(Control namingContainer)
   at System.Web.UI.Control.AddedControl(Control control, Int32 index)
   at System.Web.UI.ControlCollection.Add(Control child)
   at DotNetNuke.UI.Modules.ModuleHost.InjectModuleContent(Control content)
   at DotNetNuke.UI.Modules.ModuleHost.CreateChildControls()
   at System.Web.UI.Control.EnsureChildControls()
   at DotNetNuke.UI.Containers.Container.get_ModuleControl()
   at DotNetNuke.UI.Containers.Container.ProcessModule()
   at DotNetNuke.UI.Skins.Pane.InjectModule(ModuleInfo module)
InnerMessage:
InnerStackTrace:
Source:mscorlib
FileName:
FileLineNumber:0
FileColumnNumber:0
Method:

Awesome, thanks!  If you need an example page I can supply a URL from our customer site.  Here is the fatal error from the DNN log:

2016-10-03 17:53:53,705 [*SERVERNAME*][Thread:67][FATAL] DotNetNuke.Framework.PageBase - An error has occurred while loading page.
System.ArgumentNullException: 'dataType' argument cannot be null.
Parameter name: dataType
   at System.Data.DataColumn..ctor(String columnName, Type dataType, String expr, MappingType type)
   at System.Data.DataColumn..ctor(String columnName, Type dataType)
   at DotNetNuke.Common.Globals.BuildCrossTabDataSet(String DataSetName, IDataReader result, String FixedColumns, String VariableColumns, String KeyColumn, String FieldColumn, String FieldTypeColumn, String StringValueColumn, String NumericValueColumn, CultureInfo Culture)
   at DotNetNuke.Modules.UserDefinedTable.UserDefinedTableController.BuildMainDataSet(DataTable fieldsTable, IDataReader dr, Boolean rowMode)
   at DotNetNuke.Modules.UserDefinedTable.UserDefinedTableController.GetRow(Int32 userDefinedRowId, Boolean withPreRenderedValues, Boolean filterScript)
   at DotNetNuke.Modules.UserDefinedTable.EditForm.get_Data()
   at DotNetNuke.Modules.UserDefinedTable.EditForm.BuildEditForm()
   at DotNetNuke.Modules.UserDefinedTable.EditForm.OnInit(EventArgs e)
   at System.Web.UI.Control.InitRecursive(Control namingContainer)
   at System.Web.UI.Control.AddedControl(Control control, Int32 index)
   at System.Web.UI.Control.InitRecursive(Control namingContainer)
   at System.Web.UI.Control.AddedControl(Control control, Int32 index)
   at DotNetNuke.UI.Modules.ModuleHost.InjectModuleContent(Control content)
   at System.Web.UI.Control.EnsureChildControls()
   at System.Web.UI.Control.PreRenderRecursiveInternal()
   at System.Web.UI.Control.PreRenderRecursiveInternal()
   at System.Web.UI.Control.PreRenderRecursiveInternal()
   at System.Web.UI.Control.PreRenderRecursiveInternal()
   at System.Web.UI.Control.PreRenderRecursiveInternal()
   at System.Web.UI.Control.PreRenderRecursiveInternal()
   at System.Web.UI.Control.PreRenderRecursiveInternal()
   at System.Web.UI.Control.PreRenderRecursiveInternal()
   at System.Web.UI.Control.PreRenderRecursiveInternal()
   at System.Web.UI.Control.PreRenderRecursiveInternal()
   at System.Web.UI.Page.ProcessRequestMain(Boolean includeStagesBeforeAsyncPoint, Boolean includeStagesAfterAsyncPoint)
2016-10-03 17:53:53,705 [*SERVERNAME*][Thread:67][ERROR] DotNetNuke.Services.Exceptions.Exceptions - ~/Default.aspx?tabid=3881&error=%27dataType%27+argument+cannot+be+null.%0d%0aParameter+name%3a+dataType&content=0
System.ArgumentNullException: 'dataType' argument cannot be null.
Parameter name: dataType
   at System.Data.DataColumn..ctor(String columnName, Type dataType, String expr, MappingType type)
   at System.Data.DataColumn..ctor(String columnName, Type dataType)
   at DotNetNuke.Common.Globals.BuildCrossTabDataSet(String DataSetName, IDataReader result, String FixedColumns, String VariableColumns, String KeyColumn, String FieldColumn, String FieldTypeColumn, String StringValueColumn, String NumericValueColumn, CultureInfo Culture)
   at DotNetNuke.Modules.UserDefinedTable.UserDefinedTableController.BuildMainDataSet(DataTable fieldsTable, IDataReader dr, Boolean rowMode)
   at DotNetNuke.Modules.UserDefinedTable.UserDefinedTableController.GetRow(Int32 userDefinedRowId, Boolean withPreRenderedValues, Boolean filterScript)
   at DotNetNuke.Modules.UserDefinedTable.EditForm.get_Data()
   at DotNetNuke.Modules.UserDefinedTable.EditForm.BuildEditForm()
   at DotNetNuke.Modules.UserDefinedTable.EditForm.OnInit(EventArgs e)
   at System.Web.UI.Control.InitRecursive(Control namingContainer)
   at System.Web.UI.Control.AddedControl(Control control, Int32 index)
   at System.Web.UI.Control.InitRecursive(Control namingContainer)
   at System.Web.UI.Control.AddedControl(Control control, Int32 index)
   at DotNetNuke.UI.Modules.ModuleHost.InjectModuleContent(Control content)
   at System.Web.UI.Control.EnsureChildControls()
   at System.Web.UI.Control.PreRenderRecursiveInternal()
   at System.Web.UI.Control.PreRenderRecursiveInternal()
   at System.Web.UI.Control.PreRenderRecursiveInternal()
   at System.Web.UI.Control.PreRenderRecursiveInternal()
   at System.Web.UI.Control.PreRenderRecursiveInternal()
   at System.Web.UI.Control.PreRenderRecursiveInternal()
   at System.Web.UI.Control.PreRenderRecursiveInternal()
   at System.Web.UI.Control.PreRenderRecursiveInternal()
   at System.Web.UI.Control.PreRenderRecursiveInternal()
   at System.Web.UI.Control.PreRenderRecursiveInternal()
   at System.Web.UI.Page.ProcessRequestMain(Boolean includeStagesBeforeAsyncPoint, Boolean includeStagesAfterAsyncPoint)

--- a/App_LocalResources/FieldEditor.ascx.resx
+++ b/App_LocalResources/FieldEditor.ascx.resx
@@ -112,10 +112,10 @@
     <value>2.0</value>
   </resheader>
   <resheader name="reader">
-    <value>System.Resources.ResXResourceReader, System.Windows.Forms, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+    <value>System.Resources.ResXResourceReader, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
   </resheader>
   <resheader name="writer">
-    <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+    <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
   </resheader>
   <data name="ControlHelpText.Help" xml:space="preserve">
     <value>This is the Help Text displayed to the user to explain this field.</value>
@@ -257,5 +257,8 @@
   </data>
   <data name="ListSettings.Text" xml:space="preserve">
     <value>List Settings</value>
+  </data>
+  <data name="TitleCannotHavePipe.Text" xml:space="preserve">
+    <value>The pipe character is not allowed in the field title.</value>
   </data>
 </root>

--- a/Controls/Field.ascx
+++ b/Controls/Field.ascx
@@ -22,6 +22,17 @@
         <span class="dnnFormMessage dnnFormError" runat="server" Visible="false" id="divError"/>  
         <asp:RequiredFieldValidator  runat="server" ErrorMessage="RequiredFieldValidator" 
                                      ControlToValidate="txtFieldTitle"  resourcekey="TitleIsRequired" CssClass="dnnFormMessage dnnFormError"/>
+        <asp:CustomValidator 
+            ID="valNoPipesInTitle"            
+            runat="server" 
+            ErrorMessage="The title may not contain the pipe ( | ) character" 
+            ControlToValidate="txtFieldTitle" 
+            resourcekey="TitleCannotHavePipe" 
+            CssClass="dnnFormMessage dnnFormError" 
+            OnServerValidate="valNoPipesInTitle_ServerValidate" 
+            ClientValidationFunction="validateNoPipes" 
+            EnableClientScript="true" 
+            />
         
     </div>
     <div class="dnnFormItem" id="panHelpText" runat="server">
@@ -145,5 +156,14 @@
     function OpenHelpWindow(url) {
         window.open(url, '', 'width=800, height=800, location=no, menubar=no, resizable=yes, scrollbars=yes, status=no, toolbar=no');
         window.event.returnValue = false;
+    }
+
+    function validateNoPipes(sender, args) {
+        var text = args.Value;
+        if (text.indexOf('|')>=0) {
+            args.IsValid = false;
+            return;
+        }
+        args.IsValid = true;
     }
 </script>

--- a/Controls/Field.ascx.cs
+++ b/Controls/Field.ascx.cs
@@ -426,5 +426,16 @@ namespace DotNetNuke.Modules.UserDefinedTable.Controls
             panValidationMessage.Visible = selectedType.SupportsValidation;
             panHelpText.Visible = selectedType.SupportsEditing;
         }
+
+        protected void valNoPipesInTitle_ServerValidate(object source, System.Web.UI.WebControls.ServerValidateEventArgs args)
+        {
+            var text = args.Value;
+            if (text.Contains('|'))
+            {
+                args.IsValid = false;
+                return;
+            }
+            args.IsValid = true;
+        }
     }
 }

--- a/Controls/Field.ascx.designer.cs
+++ b/Controls/Field.ascx.designer.cs
@@ -67,6 +67,15 @@ namespace DotNetNuke.Modules.UserDefinedTable.Controls {
         protected global::System.Web.UI.HtmlControls.HtmlGenericControl divError;
         
         /// <summary>
+        /// valNoPipesInTitle control.
+        /// </summary>
+        /// <remarks>
+        /// Auto-generated field.
+        /// To modify move field declaration from designer file to code-behind file.
+        /// </remarks>
+        protected global::System.Web.UI.WebControls.CustomValidator valNoPipesInTitle;
+        
+        /// <summary>
         /// panHelpText control.
         /// </summary>
         /// <remarks>

--- a/DotNetNuke.Modules.UserDefinedTable.xml
+++ b/DotNetNuke.Modules.UserDefinedTable.xml
@@ -207,6 +207,15 @@
             To modify move field declaration from designer file to code-behind file.
             </remarks>
         </member>
+        <member name="F:DotNetNuke.Modules.UserDefinedTable.Controls.Field.valNoPipesInTitle">
+            <summary>
+            valNoPipesInTitle control.
+            </summary>
+            <remarks>
+            Auto-generated field.
+            To modify move field declaration from designer file to code-behind file.
+            </remarks>
+        </member>
         <member name="F:DotNetNuke.Modules.UserDefinedTable.Controls.Field.panHelpText">
             <summary>
             panHelpText control.

--- a/releasenotes.htm
+++ b/releasenotes.htm
@@ -5,6 +5,7 @@
     <li>Build process updated to use Nuget packages, module can now be built on vanilla Visual Studio setup even without Dnn present.</li>
     <li>Fixes and issue with the Dnn captcha not showing, adds support (optionnaly) to replace Dnn Captcha by Google ReCatcha.(Issues 21)</li>
     <li>Fixed an issue where it was impossible to apply a template created by one module into another module (Issue 26)</li>
+    <li>Prevents creating fields with a pipe ( | ) in the title, which made the module break (Issue 9)</li>
 </ul>
 
 <h2>DNN Form and List 06.04.00</h2>

